src/Service/ServiceUtils.cpp

```diff
@@ -280,149 +278,116 @@ static ErrorMessageOr<fs::path> FindSymbolsFilePathInStructuredDebugStore(
+  // 1. Create object file for the module and check if it contains symbols itself.
+  auto object_file_or_error = CreateObjectFile(module_path);
+  if (object_file_or_error.has_error()) {
+    return ErrorMessage(absl::StrFormat("Unable to create object file: %s",
+                                        object_file_or_error.error().message()));
   }
-  std::string error_message_for_client{absl::StrFormat(
-      "Unable to find debug symbols on the instance for module \"%s\". ", module_path)};
-  return ErrorMessage(error_message_for_client);
-}
-
-static ErrorMessageOr<fs::path> FindSymbolsFilePathElf(
-    const fs::path& module_path, const ElfFile* module_elf_file,
-    const std::vector<fs::path>& search_directories) {
-  CHECK(module_elf_file != nullptr);
-
-  if (module_elf_file->HasDebugSymbols()) {
+  if (object_file_or_error.value()->HasDebugSymbols()) {
     return module_path;
   }
-
-  if (module_elf_file->GetBuildId().empty()) {
-    return ErrorMessage(absl::StrFormat(
-        "Unable to find symbols for module \"%s\". Module does not contain a build id",
-        module_path));
+  // 2. If module does not contain a build id, no searching for separate symbol files can be done
+  if (object_file_or_error.value()->GetBuildId().empty()) {
+    return ErrorMessage{
+        absl::StrFormat("Module \"%s\" does not contain symbols and does not contain a build id, "
+                        "therefore Orbit cannot search for a separate symbols file",
+                        module_path.string())};
   }
-
-  std::string build_id = module_elf_file->GetBuildId();
-
+  const std::string& build_id = object_file_or_error.value()->GetBuildId();
   if (build_id.size() < 3) {
-    return ErrorMessage{
-        absl::StrFormat("The build id of \"%s\" is malformed: %s", module_path, build_id)};
+    return ErrorMessage{absl::StrFormat("The build id of \"%s\" is malformed, build id: %s",
+                                        module_path.string(), build_id)};
   }
-
+  // 3. Search in structured symbols stores.
   const fs::path structured_debug_store{"/usr/lib/debug"};
   ErrorMessageOr<fs::path> debug_store_result =
       FindSymbolsFilePathInStructuredDebugStore(structured_debug_store, build_id);
   if (debug_store_result.has_value()) return debug_store_result.value();
 
+  // 4. Search in hardcoded directories + additional directoried (user provided) + next to the
+  // module
+  std::vector<fs::path> search_directories{kHardcodedSearchDirectories};
+  const auto& additional_search_directories = request.additional_search_directories();
+  search_directories.insert(search_directories.end(), additional_search_directories.begin(),
+                            additional_search_directories.end());
+  search_directories.push_back(object_file_or_error.value()->GetFilePath().parent_path());
+
   const fs::path& filename = module_path.filename();
   fs::path filename_dot_debug = filename;
   filename_dot_debug.replace_extension(".debug");
```

Comment:
Should the .debug / .pdb filenames only be added to the search paths in the ELF / COFF case, respectively? O/w could we have weird name clashes with finding a pdb for an ELF file accidentally? We would certainly bail out when we check the build id, but that could still lead to weird errors at least.

Reply:
I also think thats a good idea to only search for .pdb when coff (and .debug when elf). I added this here (OrbitService) and also in SymbolHelper (Client).